multi/multi.go

```diff
@@ -0,0 +1,271 @@
+package multi
+
+import (
+	"errors"
+	"github.com/tarantool/go-tarantool"
+	"log"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	connDisconnected = iota
+	connConnected
+	connClosed
+)
+
+type ConnectionMulti struct {
+	addrs    []string
+	connOpts tarantool.Opts
+	opts     OptsMulti
+
+	mutex           sync.RWMutex
+	notify			chan tarantool.ConnEvent
+	state           uint32
+	control			chan struct{}
+	currentPoolAddr string
+	pool            map[string]*tarantool.Connection
+}
+
+var _ = tarantool.Connector(&ConnectionMulti{}) // check compatibility with connector interface
+
+type OptsMulti struct {
+	Chooser ConnectionChooser
+}
+
+func ConnectWithOpts(addrs []string, connOpts tarantool.Opts, opts OptsMulti) (connMulti *ConnectionMulti, err error) {
+	if len(addrs) == 0 {
+		return nil, errors.New("addrs should not be empty")
+	}
+	if opts.Chooser == nil {
+		return nil, errors.New("OptsMulti.Chooser should not be nil")
+	}
+
+	notify := make(chan tarantool.ConnEvent)
+	connOpts.Notify = notify
+	connMulti = &ConnectionMulti{
+		addrs:    addrs,
+		connOpts: connOpts,
+		opts:     opts,
+		notify:	  notify,
+		control:  make(chan struct{}),
+		pool:     make(map[string]*tarantool.Connection),
+	}
+
+	if _, err := connMulti.getNextConnection(false); err != nil {
+		return nil, err
+	}
+
+	go connMulti.checker()
+
+	return connMulti, nil
+}
+
+func Connect(addrs []string, connOpts tarantool.Opts) (connMulti *ConnectionMulti, err error) {
+	opts := OptsMulti{
+		Chooser: &RoundRobinChooser{},
+	}
+	return ConnectWithOpts(addrs, connOpts, opts)
+}
+
+func (connMulti *ConnectionMulti) checker() {
+	for connMulti.state != connClosed {
```

Comment:
connMulti.state is written under lock using atomics. Therefore you should read it either with atomic or under lock.